hsmd/hsmd.c

```diff
@@ -735,6 +735,30 @@ static struct io_plan *init_hsm(struct io_conn *conn,
 						    &secretstuff.bip32)));
 }
 
+/*~ The libsecp256k1 EC Diffie Hellman function provides X and Y coordinates
+ * of the point.
+ * We want the point itself, but the official constructors for struct pubkey
+ * use DER encoding, so we use a temporary DER encoding in-memory and
+ * convert from that.
+ */
+static int ecdh_load_point(u8 *uoutput,
+			   const u8 *x, const u8 *y,
+			   void *data UNUSED)
+{
+	struct pubkey *output = (struct pubkey *) uoutput;
+	u8 der_encoding[PUBKEY_CMPR_LEN];
+	/* Copied from libsecp256k1 code, no idea why this is.  */
+	u8 version = (y[31] & 0x01) | 0x02;
+
+	der_encoding[0] = version;
+	memcpy(&der_encoding[1], x, PUBKEY_CMPR_LEN - 1);
+
+	if (pubkey_from_der(der_encoding, sizeof(der_encoding), output))
```

Comment:
This copies 64 bytes (in implementation-defined order!) into the "hash function" output.

I'm pretty sure that if we do point pass-through like this, ecdh can be made to reveal the secret key by passing an x point of "1".

IIRC this was the entire rationale behind the hashing before ECDH.  I am not comfortable discarding it.

Reply:
> This copies 64 bytes (in implementation-defined order!) into the "hash function" output.

This is true, but it really is a struct pubkey, just passed through secp256k1_ecdh in a u8 *.  It's fine.

> I'm pretty sure that if we do point pass-through like this, ecdh can be made to reveal the secret key by passing an x point of "1".
> 
> IIRC this was the entire rationale behind the hashing before ECDH. I am not comfortable discarding it.

Nope, this hashing is on the output, not the input.  It's fine, though a bit weird, yes.